### PR TITLE
Added strand sort algorithm

### PR DESCRIPTION
**Pull request by mertcandav**
#### Description of Change
Added strand sort algorithm.

#### Checklist
- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1081"></a>

### CODE REVIEW

**Review by Panquesito7 — COMMENTED** [review]
This seems to be more like a C program, rather than a C++ program.
Would you like this algorithm to be implemented in the [C](https://github.com/TheAlgorithms/C) repository?

If you still want this in C++, we can improve
it further and make it a C++ program. 🙂

**Comment by mertcandav** [comment]
Yes it can be implemented, my goal is just to contribute :)

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please move this file to the `sorting` directory.

**Comment by mertcandav** [comment]
Moved to ``sorting`` directory.

File: sorting/strand_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#include <iostream>
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:1]: See #962 as a reference on how to document the first comment block.
```suggestion
/**
 * @file
 * @brief description of algorithm
 *
 * @details
 * <full-description here>
 *
 * @author [Mertcan Davulcu](https://github.com/mertcandav)
 */
#include <iostream>
```

File: sorting/strand_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <list>
+
+template <typename T>
+std::list<T> strand_sort(std::list<T> lst) {
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:5]: Missing function documentation, parameter
documentation and return statement documentation.

See #962 as a reference.

File: sorting/strand_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            } else {
+                it++;
+            }
+        }
+        result.merge(sorted);
+    }
+    return result;
+}
+
+// Entry point for testing
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:27]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/strand_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * + Now that the sorted array is not empty, the newly extracted substring is merged with the sorted array.
+ * + Repeat types 3 and 4 until the sub-sequence and main sequence are empty.
+ * 
+ * @author [Mertcan Davulcu](https://github.com/mertcandav)
+ */
+#include <iostream>
+#include <iostream>
+#include <list>
+
+namespace strand {
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:21]:
```suggestion
/**
 * @namespace sorting
 * @brief Sorting algorithms
 */
namespace sorting {
/**
 * @namespace strand
 * @brief Functions for [Strand Sort](https://en.wikipedia.org/wiki/Strand_sort) algorithm
 */
namespace strand {
```

File: sorting/strand_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                    it = lst.erase(it); // Deletes the element with the current iterator and assigns the deleted element to the iterator.
+                } else {
+                   it++; // Next iterator.
+                }
+            }
+            result.merge(sorted); // Merge "result" list with "sorted" list.
+          }
+        return result; // Returns sorted list
+    }
+} // namespace strand
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:49]:
```suggestion
}  // namespace strand
}  // namespace sorting
```

File: sorting/strand_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#include <iostream>
+#include <list>
+
+namespace strand {
+    // Element type of list.
+    template <typename T>
+    /*
+    * Apply sorting.
+    * lst: List to sorted.
+    */
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:27]: You can also provide further information by using
`@details` as in the first comment block. 🙂

See #962 as a reference for how to document functions.
```suggestion
    /**
    * @brief Apply sorting
    * @param lst List to be sorted
    * @returns <provide description here>
    */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/strand_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * @namespace sorting
+ * @brief Sorting algorithms
+ */
+namespace sorting {
+    /**
+    * @namespace strand
+    * @brief Functions for [Strand Sort](https://en.wikipedia.org/wiki/Strand_sort) algorithm
+    */
+    namespace strand {
+        // Element type of list.
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:31]:
```suggestion
```

File: sorting/strand_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        template <typename T>
+        /**
+        * @brief Apply sorting
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:34]:
```suggestion
        * @brief Apply sorting
        * @tparam element type of list
```

File: sorting/strand_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    */
+    namespace strand {
+        template <typename T>
+        /**
+        * @brief Apply sorting
+        * @tparam element type of list
+        * @param lst List to be sorted
+        * @returns Sorted list<T> instance
+        */
+        std::list<T> strand_sort(std::list<T> lst) {
```
Comment (Panquesito7) [inline sorting/strand_sort.cpp:38]:
```suggestion
        /**
        * @brief Apply sorting
        * @tparam element type of list
        * @param lst List to be sorted
        * @returns Sorted list<T> instance
        */
        template <typename T>
        std::list<T> strand_sort(std::list<T> lst) {
```

**Review by Panquesito7 — APPROVED** [review]
LGTM, good work. 👍

File: sorting/strand_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
+/**
+ * @file strand_sort.cpp
+ * @brief Implementation of [Strand Sort](https://en.wikipedia.org/wiki/Strand_sort) algorithm.
+ *
+ * @details
+ * Strand Sort is a sorting algorithm that works in (O(n)) time if list is already sorted and works in (O(n * n)) in worst case.
```
Comment (kvedala) [inline sorting/strand_sort.cpp:6]:
```suggestion
 * Strand Sort is a sorting algorithm that works in \f$O(n)\f$ time if list is already sorted and works in \f$O(n^2)\f$ in worst case.
```

File: sorting/strand_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+ * 
+ * ### Algorithm
+ * + It is passed over the array to be sorted once and the ascending (sequential) numbers are taken.
```
Comment (kvedala) [inline sorting/strand_sort.cpp:9]:
```suggestion
 * It is passed over the array to be sorted once and the ascending (sequential) numbers are taken.
```
1. please remove the `### Algorithm` as it is redundant and will be properly formatted by doxygen
2. please remove the preceding `+` on the lines. If implying bullet points, then use `*` or `-`

File: sorting/strand_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+ * + After the first iteration, the sequential sub-array is put on the empty sorted array.
+ * + The main sequence is passed over again and a new sub-sequence is created in order.
+ * + Now that the sorted array is not empty, the newly extracted substring is merged with the sorted array.
+ * + Repeat types 3 and 4 until the sub-sequence and main sequence are empty.
+ * 
+ * @author [Mertcan Davulcu](https://github.com/mertcandav)
+ */
+#include <iostream>
+#include <iostream>
```
Comment (kvedala) [inline sorting/strand_sort.cpp:18]: duplicate line
```suggestion
```

File: sorting/strand_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                    }
+                }
+                result.merge(sorted); // Merge "result" list with "sorted" list.
+            }
+            return result; // Returns sorted list
+        }
+    }  // namespace strand
+}  // namespace sorting
+
+/**
```
Comment (kvedala) [inline sorting/strand_sort.cpp:62]: please add a `static void test() {}` function. as in other sorting algorithm programs. 
https://github.com/TheAlgorithms/C-Plus-Plus/blob/d4b54b8294722607f8ed1e29f2b3ff500ba787e7/sorting/merge_insertion_sort.cpp#L128

**Review by kvedala — CHANGES_REQUESTED** [review]
well documented code 👍

**Comment by kvedala** [comment]
Awesome. Can you enable GitHub Actions using this link: https://github.com/mertcandav/C-Plus-Plus/actions
Any minor commit will trigger the actions on your branch. This will ensure any syntax corrections and additional commits are managed appropriately.

**Comment by mertcandav** [comment]
I think I activated it.

**Comment by kvedala** [comment]
> I think I activated it.

Please make a small commit to trigger the actions. For example, delete line# 8 in the code. It is unnecessary. 
If the actions have been enabled, we should see a commit made by the GitHub-actions to the file DIRECTORY.md.

**Comment by mertcandav** [comment]
Yes, enabled.

**Review by kvedala — APPROVED** [review]
LGTM 👍

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**